src/renderer/atlas/shader_ps.hlsl

```diff
@@ -100,45 +95,69 @@ float4 main(float4 pos: SV_Position): SV_Target
+        {
+            // The cursor texture is stored at the top-left-most glyph cell.
+            // Cursor pixels are either entirely transparent or opaque.
+            // --> We can just use .a as a mask to flip cursor pixels on or off.
+            color = alphaBlendPremultiplied(color, decodeRGBA(cursorColor) * glyphs[cellPos].a);
+        }
     }
 
     // Layer 2:
     // Step 1: Underlines
-    if ((cell.flags & CellFlags_Underline) && cellPos.y >= underlinePos.x && cellPos.y < underlinePos.y)
+    [branch] if (cell.flags & CellFlags_Underline)
     {
-        color = alphaBlendPremultiplied(color, fg);
+        [flatten] if (cellPos.y >= underlinePos.x && cellPos.y < underlinePos.y)
+        {
+            color = alphaBlendPremultiplied(color, fg);
+        }
     }
-    if ((cell.flags & CellFlags_UnderlineDotted) && cellPos.y >= underlinePos.x && cellPos.y < underlinePos.y && (viewportPos.x / (underlinePos.y - underlinePos.x) & 3) == 0)
+    [branch] if (cell.flags & CellFlags_UnderlineDotted)
     {
-        color = alphaBlendPremultiplied(color, fg);
+        [flatten] if (cellPos.y >= underlinePos.x && cellPos.y < underlinePos.y && (viewportPos.x / (underlinePos.y - underlinePos.x) & 3) == 0)
+        {
+            color = alphaBlendPremultiplied(color, fg);
+        }
     }
     // Step 2: The cell's glyph, potentially drawn in the foreground color
     {
         float4 glyph = glyphs[decodeU16x2(cell.glyphPos) + cellPos];
 
-        if (cell.flags & CellFlags_ColoredGlyph)
+        [branch] if (cell.flags & CellFlags_ColoredGlyph)
         {
             color = alphaBlendPremultiplied(color, glyph);
         }
-        else if (useClearType)
-        {
-            color = DWrite_CleartypeBlend(gammaRatios, enhancedContrast, false, color, fg, glyph);
-        }
         else
         {
-            color = alphaBlendPremultiplied(color, DWrite_GrayscaleBlend(gammaRatios, enhancedContrast, false, fg, glyph.a));
+            float3 foregroundStraight = DWrite_UnpremultiplyColor(fg);
+            float blendEnhancedContrast = DWrite_ApplyLightOnDarkContrastAdjustment(enhancedContrast, foregroundStraight);
+
+            [branch] if (useClearType)
+            {
+                // See DWrite_ClearTypeBlend
+                float3 contrasted = DWrite_EnhanceContrast3(glyph.rgb, blendEnhancedContrast);
+                float3 alphaCorrected = DWrite_ApplyAlphaCorrection3(contrasted, foregroundStraight, gammaRatios);
+                color = float4(lerp(color.rgb, foregroundStraight, alphaCorrected * fg.a), 1.0f);
+            }
+            else
+            {
+                // See DWrite_GrayscaleBlend
+                float intensity = DWrite_CalcColorIntensity(foregroundStraight);
+                float contrasted = DWrite_EnhanceContrast(glyph.a, blendEnhancedContrast);
+                color = fg * DWrite_ApplyAlphaCorrection(contrasted, intensity, gammaRatios);
+            }

```

Comment:
I inlined the two DWrite functions to offset the binary size cost for the shader a bit. Due to all the `[branch]` annotations the compiler can't inline as much which increases binary size of the shader by about 50%.